collector/adcs.go

```diff
@@ -0,0 +1,246 @@
+//go:build windows
+// +build windows
+
+package collector
+
+import (
+	"errors"
+	"github.com/prometheus-community/windows_exporter/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"strings"
+)
+
+func init() {
+	registerCollector("adcs", ADCSCollectorMethod, "Certification Authority")
+}
+
+type ADCSCollector struct {
+	RequestsPerSecond                            *prometheus.Desc
+	RequestProcessingTime                        *prometheus.Desc
+	RetrievalsPerSecond                          *prometheus.Desc
+	RetrievalProcessingTime                      *prometheus.Desc
+	FailedRequestsPerSecond                      *prometheus.Desc
+	IssuedRequestsPerSecond                      *prometheus.Desc
+	PendingRequestsPerSecond                     *prometheus.Desc
+	RequestCryptographicSigningTime              *prometheus.Desc
+	RequestPolicyModuleProcessingTime            *prometheus.Desc
+	ChallengeResponsesPerSecond                  *prometheus.Desc
+	ChallengeResponseProcessingTime              *prometheus.Desc
+	SignedCertificateTimestampListsPerSecond     *prometheus.Desc
+	SignedCertificateTimestampListProcessingTime *prometheus.Desc
+}
+
+// ADCSCollectorMethod ...
+func ADCSCollectorMethod() (Collector, error) {
+	const subsystem = "adcs"
+	return &ADCSCollector{
+		RequestsPerSecond: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "requests_total"),
+			"Certificate requests processed per second",
+			[]string{"cert_template"},
+			nil,
+		),
+		RequestProcessingTime: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "request_processing_time_seconds"),
+			"Time elapsed per request",
+			[]string{"cert_template"},
+			nil,
+		),
```

Comment:
Does this metric represent the total request processing time for a template, or the average time spent on each request?

Reply:
It is the *last* reported value by Windows for each template. Have updated the metric description to read "Last time elapsed for..."

Reply:
To further clarify, there is a label of `_Total` which would report the *last* known time across *all* certificate templates.
```
# HELP windows_adcs_request_processing_time Time elapsed per request
# TYPE windows_adcs_request_processing_time gauge
windows_adcs_request_processing_time{cert_template="Administrator"} 0
windows_adcs_request_processing_time{cert_template="CEPEncryption"} 0
windows_adcs_request_processing_time{cert_template="DirectoryEmailReplication"} 0
windows_adcs_request_processing_time{cert_template="DomainController"} 11.294
windows_adcs_request_processing_time{cert_template="DomainControllerAuthentication"} 0
windows_adcs_request_processing_time{cert_template="EFS"} 0
windows_adcs_request_processing_time{cert_template="EFSRecovery"} 0
windows_adcs_request_processing_time{cert_template="EnrollmentAgentOffline"} 0
windows_adcs_request_processing_time{cert_template="IPSECIntermediateOffline"} 0
windows_adcs_request_processing_time{cert_template="KerberosAuthentication"} 11.419
windows_adcs_request_processing_time{cert_template="Machine"} 0
windows_adcs_request_processing_time{cert_template="SubCA"} 0
windows_adcs_request_processing_time{cert_template="User"} 0
windows_adcs_request_processing_time{cert_template="WebServer"} 0
windows_adcs_request_processing_time{cert_template="Workstation"} 0
windows_adcs_request_processing_time{cert_template="_Total"} 11.294
```

Reply:
I think this is representing the total time spent processing requests, rather than the total time of the most recent request.

Does this metric increment over time, or increase/decrease with each reading?

Reply:
I confirmed using Performance Monitor that the metric reports the time for the last request, and therefore can increase/decrease as requests are processed by ADCS. I believe the screenshot I attached below shows this.

Reply:
Ah, don't mind me then. I had assumed it was a counter of some kind, which is common with the perflib metrics.